### PR TITLE
Ingress Documentation

### PR DESCRIPTION
**Pull request by louiseschmidtgen**
## Description
This PR adds docs around Ingress.
Have a look at: https://canonical-ubuntu-documentation-library--376.com.readthedocs.build/canonical-kubernetes/376/snap/explanation/ingress/

### CODE REVIEW

File: docs/src/snap/explanation/ingress.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+## Kubernetes Pods and Services
+
+In Kubernetes, the smallest unit is a pod, which encapsulates application containers.
+Since pods are ephemeral and their IP addresses change when destroyed and restarted,
+they are exposed through services.
+Services offer a stable network interface by providing discoverable names and
+load balancing functionality for managing a set of pods.
+For further details on Kubernetes Services,
+refer to the upstream documentation: [Kubernetes Service][Service].
```
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:16]:
```suggestion
refer to the [upstream Kubernetes Service documentation][Service].
```

File: docs/src/snap/explanation/ingress.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+[Ingress Rules][Ingress Rules].
+
+While the Ingress Resource manages the routing rules for the incoming traffic,
+the [Ingress Controller][Ingress Controller] is responsible for implementing
+those rules by configuring the underlying networking infrastructure of the cluster.
+Ingress does not work without an Ingress Controller.
+
+The Ingress Controller also serves as a layer 7 (HTTP/HTTPS) load balancer
+that routes traffic from outside of your cluster to services inside of your cluster.
+Please do not confuse this with the Kubernetes Service LoadBalancer type.
```
Comment (bschimke95) [inline docs/src/snap/explanation/ingress.md:37]: Maybe add a brief sentence about the difference between both.

File: docs/src/snap/explanation/ingress.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+```{kroki} ../../assets/ingress.puml
+```
+
+With Canonical Kubernetes, enabling default Ingress is easy:
+See [default Ingress][Ingress].
+At the moment of writing, enabling default Ingress will create a
```
Comment (bschimke95) [inline docs/src/snap/explanation/ingress.md:44]: This sounds a bit like "at the moment we use Cilium but that might change at any moment". Maybe just remove the specific type altogether.

File: docs/src/snap/explanation/ingress.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+[Cilium Ingress Controller][Cilium Ingress Controller] for you.
+If your cluster requires different Ingress Controllers,
+the responsibility of implementation falls upon you.
+
+You will need to create the Ingress Resource,
+outlining rules that direct traffic to your application's Kubernetes service.
+
+## Network
```
Comment (bschimke95) [inline docs/src/snap/explanation/ingress.md:52]: Maybe put this at the beginning. This reads like a prerequisite.

**Review by bschimke95 — COMMENTED** [review]
nice read, added some comments.

File: docs/src/snap/explanation/ingress.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+Since upstream Kubernetes comes without a network provider,
+it requires the use of a [network plugin][network plugin].
+This network plugin facilitates communication between pods,
+services, and external resources, ensuring smooth traffic flow within the cluster.
+The current implementation of Canonical Kubernetes leverages a widely adopted
+CNI (Container Network Interface) called [Cilium][Cilium].
+If you wish to use a different network plugin
+the implementation and configuration falls under your responsibility.
+
+Learn how to use the Canonical Kubernetes default network: See [Network][Network].
```
Comment (bschimke95) [inline docs/src/snap/explanation/ingress.md:67]: Integrate the link into the sentence as Nick suggested in another place.
```
Learn how to use the Canonical Kubernetes default network in the [networking HowTo][Network]
```

File: docs/src/snap/explanation/ingress.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+Since pods are ephemeral and their IP addresses change when destroyed and restarted,
+they are exposed through services.
+Services offer a stable network interface by providing discoverable names and
+load balancing functionality for managing a set of pods.
+For further details on Kubernetes Services,
+refer to the upstream documentation: [Kubernetes Service][Service].
+
+## Ingress
+
+[Ingress][Ingress K8s] is a Kubernetes Resource that manages
```
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:20]: is that 'R' really an 'r'?

File: docs/src/snap/explanation/ingress.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+external access by handling both HTTP and HTTPS traffic to services within your cluster.
+Traffic routed through the Ingress is directed to a service,
+which in turn forwards it to the relevant pod
+running the desired application within a container.
+
+The Ingress Resource lets you define rules on how traffic should get handled.
+Learn more about these rules and how to implement them here:
```
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:27]: probably better to say "Refer to the [Kubernetes documentation on Ingress rules] for up to date information on the available rules and their implementation." or something like that.

File: docs/src/snap/explanation/ingress.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+The Ingress Controller also serves as a layer 7 (HTTP/HTTPS) load balancer
+that routes traffic from outside of your cluster to services inside of your cluster.
+Please do not confuse this with the Kubernetes Service LoadBalancer type.
+
+```{kroki} ../../assets/ingress.puml
+```
+
+With Canonical Kubernetes, enabling default Ingress is easy:
+See [default Ingress][Ingress].
```
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:43]:
```suggestion
See the [default Ingress guide][Ingress].
```

File: docs/src/snap/explanation/ingress.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+At the moment of writing, enabling default Ingress will create a
+[Cilium Ingress Controller][Cilium Ingress Controller] for you.
+If your cluster requires different Ingress Controllers,
+the responsibility of implementation falls upon you.
+
```
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:48]: I think it would be useful to add a note here which more specifically calls out that the underlying mechanism  provided by default will always operate in the same way but may not always be Cilium (if indeed, that is accurate!)

File: docs/src/snap/explanation/ingress.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+You will need to create the Ingress Resource,
+outlining rules that direct traffic to your application's Kubernetes service.
+
+## Network
+
+In order to use Canonical Kubernetes default Ingress ensure that the
```
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:54]: I wonder if there is any way we can get around repeating Canonical Kubernetes quite so often? And the word 'default'. 
Maybe reword the whole paragraph

When you install Canonical Kubernetes, the default Network is automatically enabled. This is also a requirement for the default Ingress to function.

File: docs/src/snap/explanation/ingress.md (hunk 5 above)
Comment (evilnick) [inline docs/src/snap/explanation/ingress.md:67]: Refer to the [default Network guide] for more details

**Review by evilnick — CHANGES_REQUESTED** [review]
Nice work! Just a few of my usual comments

File: docs/src/snap/explanation/ingress.md (hunk 4 above)
Comment (louiseschmidtgen) [reply docs/src/snap/explanation/ingress.md]: Sure thing!

File: docs/src/snap/explanation/ingress.md (hunk 9 above)
Comment (louiseschmidtgen) [reply docs/src/snap/explanation/ingress.md:70]: So this one is a bit of a slippery slope. If we talk about it it may sound like we may change it some time. Have a look at my change and see what you think. @bschimke95 @evilnick you might have a different opinion?

File: docs/src/snap/explanation/ingress.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
+# Ingress
+
+In Kubernetes, understanding how inbound traffic is managed inside of your
+cluster can be complex.
+While there is abundant documentation, this explanation provides you with the essentials
```
Comment (eaudetcobello) [inline docs/src/snap/explanation/ingress.md:5]: Can you mention where? Or rephrase maybe

File: docs/src/snap/explanation/ingress.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+While the Ingress resource manages the routing rules for the incoming traffic,
+the [Ingress Controller][Ingress Controller] is responsible for implementing
+those rules by configuring the underlying networking infrastructure of the cluster.
+Ingress does not work without an Ingress Controller.
+
+The Ingress Controller also serves as a layer 7 (HTTP/HTTPS) load balancer
+that routes traffic from outside of your cluster to services inside of your cluster.
+Please do not confuse this with the Kubernetes Service LoadBalancer type
+which operates at layer 4 and routes traffic directly to individual pods.
```
Comment (eaudetcobello) [inline docs/src/snap/explanation/ingress.md:55]: Nice

**Review by eaudetcobello — APPROVED** [review]
Nice work!

File: docs/src/snap/explanation/ingress.md (hunk 9 above)
Comment (evilnick) [reply docs/src/snap/explanation/ingress.md:70]: "We reserve the right to..." sounds more like we are definitely planning to change it :joy: 

perhaps we should ask @perk

File: docs/src/snap/explanation/ingress.md [hunk 13]
```diff
@@ -13,59 +29,48 @@ they are exposed through services.
...
 With Canonical Kubernetes, enabling default Ingress is easy:
-See [default Ingress][Ingress].
-At the moment of writing, enabling default Ingress will create a
-[Cilium Ingress Controller][Cilium Ingress Controller] for you.
+See the [default Ingress guide][Ingress].
+Enabling default Ingress will create an
+[Ingress Controller][Cilium Ingress Controller] for you.
+The underlying mechanism provided by default will always operate in the same way,
+and will be Cilium for the foreseeable future.
+However, we do reserve the right to change it in the future if necessary.
```
Comment (perk) [inline docs/src/snap/explanation/ingress.md:66]:
```suggestion
The underlying mechanism provided by default is currently Cilium. However, it should always be operated through the provided CLI rather than directly. This way, we can provide the best experience for future cluster maintenance and upgrades.
```

File: docs/src/snap/explanation/ingress.md (hunk 9 above)
Comment (perk) [reply docs/src/snap/explanation/ingress.md:70]: I propose the following wording here:
> The underlying mechanism provided by default is currently Cilium. However, it should always be operated through the provided CLI rather than directly. This way, we can provide the best experience for future cluster maintenance and upgrades.

See my suggestion here: https://github.com/canonical/k8s-snap/pull/376/files/76cde6aa0ac7e6d57ead5a8938e0e0fda0deb537..c585437d88e2da9c006e215369d333bb6cf9395f#r1579796674

File: docs/src/snap/explanation/ingress.md (hunk 13 above)
Comment (louiseschmidtgen) [reply docs/src/snap/explanation/ingress.md:66]: Much better put, thanks! @perk

File: docs/src/snap/explanation/ingress.md (hunk 9 above)
Comment (louiseschmidtgen) [reply docs/src/snap/explanation/ingress.md:71]: Thank you!

**Comment by evilnick** [comment]
🍰

**Review by evilnick — APPROVED** [review]
